### DISCUSSION TITLE
How to use server code from a ClientComponent?

### DISCUSSION THREAD

**Discussion by fahadahmed** [vercel/next.js#44415] [Help] [answered]
## Overview
I am creating a user authentication workflow using the latest `appDir` in Next13 via Firebase Authentication

## Issue
When trying to verify the idToken I get from the client side component, I try and use the firebase admin auth to verify the user token. However, in order the LoginForm component is marked as client component via `'use client'` convention in order to get the component to get a user token after the form has been submitted via an `onClick` event.

```ts
// LoginForm.tsx
import React from 'react';
import { signInWithEmailAndPassword } from 'firebase/auth';
import { auth } from '../../lib/firebaseClient';
import { createUserSession } from '../../lib/session';

export default function LoginForm() {
  const handleLogin = async () => {
    const email = 'fahad@email.com';
    const password = 'xxx';
    const { user } = await signInWithEmailAndPassword(auth, email, password);
    const token = await user.getIdToken();
    console.log('token', token)
    createUserSession(token, '/admin');
  }
  return (
    <div>
      <h3>Login</h3>
      <input type="text" name="email" placeholder="email" />
      <input type="password" name="password" placeholder="password" />
      <button type="submit" onClick={handleLogin}>Login</button>
    </div>
  )
};
```
I have a `session.ts` file that needs to take care of generating a session cookie. Some of the code is here:

```ts
// sesssion.ts
async function getSessionToken(idToken: any) {
  const token = await idToken;
  try {
    const decodedToken = await admin.auth().verifyIdToken(token);
    console.log(decodedToken);
    if (new Date().getTime() / 1000 - decodedToken.auth_time > 5 * 60) {
      throw new Error('Recent sign in required');
    }
    const twoWeeks = 60 * 60 * 24 * 14 * 1000;
    return admin.auth().createSessionCookie(idToken, { expiresIn: twoWeeks });
  } catch (error: any) {
    throw new Error(error);
  }
}

export async function createUserSession(idToken: string, redirectTo: string) {
  const token = await getSessionToken(idToken);
....
// once I get the session token, I need to go on with the rest of the flow
}
```

The issue I am having is that when Firebase's `admin.auth()` is being called, I get an error: `Module not found: cannot resolve 'fs'`

![Screenshot 2022-12-29 at 2 57 44 pm](https://user-images.githubusercontent.com/1190340/209904415-99344d56-78d8-4541-b031-8d9fd6719e35.png)

I believe this is because `fs` is available on node environment and I am trying to access it inside a browser environment since `LoginForm` is a client component. 

I would like to know what is the best way to implement such a workflow.

Thanks for reading my issue.

**Comment by icyJoseph** [comment, marked as answer]
This is not necessarily an issue with client components, or the app dir, rather, simply put, you are trying to use server code client side. Which would also fail in pages. 

To tackle this you need to create an API route, and place the firebase auth logic there, POST the inputs needed to that API route, and return the session token. What's more you could do everything within the API route.

**Reply by fahadahmed** [reply]
Thanks @icyJoseph, your tips help. 

Having the server code on the API routes allowed me to have it interact within the client component via POST.
